index.ts

```diff
@@ -13,160 +41,480 @@ enum Tile {
+    removeLock1(RawTile.LOCK1);
+    moveToTile(playerx, playery + dx); 
+  }
+}
+
+class Lock1 implements Tile {
+  isAir() { return false; }
+  isFlux() { return false; }
+  isUnbreakable() { return false; }
+  isPlayer() { return false; }
+  isStone() { return false; }
+  isFallingStone() { return false; }
+  isBox() { return false; }
+  isFallingBox() { return false; }
+  isKey1() { return false; }
+  isLock1() { return true; }
+  isKey2() { return false; }
+  isLock2() { return false; }
+  isStony() { return false; }
+  isBoxy() { return false; }
+  draw(g: CanvasRenderingContext2D, x: number, y: number) { 
+    g.fillStyle = '#ffcc00';
+    g.fillRect(x * TILE_SIZE, y * TILE_SIZE, TILE_SIZE, TILE_SIZE);
+  }
+  isEdible() { return false; }
+  isPushable() { return false; }
+  moveHorizontal(dx: number) { }
+  moveVertical(dx: number) { }
+}
+
+class Key2 implements Tile {
+  isAir() { return false; }
+  isFlux() { return false; }
+  isUnbreakable() { return false; }
+  isPlayer() { return false; }
+  isStone() { return false; }
+  isFallingStone() { return false; }
+  isBox() { return false; }
+  isFallingBox() { return false; }
+  isKey1() { return false; }
+  isLock1() { return false; }
+  isKey2() { return true; }
+  isLock2() { return false; }
+  isStony() { return false; }
+  isBoxy() { return false; }
+  draw(g: CanvasRenderingContext2D, x: number, y: number) { 
+    g.fillStyle = '#00ccff';
+    g.fillRect(x * TILE_SIZE, y * TILE_SIZE, TILE_SIZE, TILE_SIZE);
+  }
+  isEdible() { return false; }
+  isPushable() { return false; }
+  moveHorizontal(dx: number) {
+    removeLock2(RawTile.LOCK2);
+    moveToTile(playerx + dx, playery);
+  }
+  moveVertical(dx: number) { 
+    removeLock2(RawTile.LOCK2);
+    moveToTile(playerx, playery + dx); 
+  }
+}
+
+class Lock2 implements Tile {
+  isAir() { return false; }
+  isFlux() { return false; }
+  isUnbreakable() { return false; }
+  isPlayer() { return false; }
+  isStone() { return false; }
+  isFallingStone() { return false; }
+  isBox() { return false; }
+  isFallingBox() { return false; }
+  isKey1() { return false; }
+  isLock1() { return false; }
+  isKey2() { return false; }
+  isLock2() { return true; }
+  isStony() { return false; }
+  isBoxy() { return false; }
+  draw(g: CanvasRenderingContext2D, x: number, y: number) { 
+    g.fillStyle = '#00ccff';
+    g.fillRect(x * TILE_SIZE, y * TILE_SIZE, TILE_SIZE, TILE_SIZE);
+  }
+  isEdible() { return false; }
+  isPushable() { return false; }
+  moveHorizontal(dx: number) { }
+  moveVertical(dx: number) { }
+}
+
+// Input Enum & Interface & Class
+enum RawInput {
   UP, DOWN, LEFT, RIGHT
 }
 
+interface Input {
+  isRight(): boolean;
+  isLeft(): boolean;
+  isUp(): boolean;
+  isDown(): boolean;
+  handle(): void;
+}
+
+class Right implements Input {
+  isRight() { return true; }
+  isLeft() { return false; }
+  isUp() { return false; }
+  isDown() { return false; }
+  handle() { moveToTile(playerx + 1, playery); }
```

Comment:
```suggestion
  handle() { moveHorizontal(1); }
```
책에서는 해당 메서드로 구현 되어 있는데 `moveToTile`로 왜 변경되었는지가 궁금합니다.

Reply:
제가 이해하고 작성하기로는 moveHorizontal 과 moveVertical 이 인터페이스에서 상속 받아 
클래스 안에 구현된 내장 메소드의 형태로 이전되었습니다.
그래서 책을 따라서 코드를 작성하면 input 이라는 Input 인터페이스를 상속 받는 클래스들은
moveHorizontal 과 moveVertical 을 호출할 수 없기 때문에 빨간줄(에러) 가 발생합니다.
moveToTile 이란 메소드는 playerx, playery 라는 좌표 전역 변수를 재설정하고 있습니다.
그래서 저 부분을 바꾸었습니다.

Reply:
moveToTile 메서드로 두 가지 일을 하는 것으로 보입니다. `moveHorizontal`와 `moveVertical`따로 구분한 이유 중 하나 일 것입니다. 상속 받은 클래스에 호출 할 수 없다고 에러가 발생한 이유가 interface에서 상속 받은 클래스에서 메서드 구현이 제대로 되어 있지 않아서 문제가 발생 할 수 있습니다.